File: backend/app/schemas/progress.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel


class ProgressEntry(BaseModel):
    month_name: str
    completed: bool
    completed_at: Optional[datetime] = None


class ProgressUpdateRequest(BaseModel):
    roadmap_id: str
    month_name: str
    completed: bool


class ProgressUpdateResponse(BaseModel):
    id: str
    user_id: str
    roadmap_id: str
    month_name: str
    completed: bool
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ProgressListResponse(BaseModel):
    entries: list[ProgressEntry]
    total_months: int
    completed_months: int
    percentage: int

File: frontend/src/lib/api/progress.ts
import { api } from './client'

export interface ProgressEntry {
  month_name: string
  completed: boolean
  completed_at: string | null
}

export interface ProgressSummary {
  entries: ProgressEntry[]
  total_months: number
  completed_months: number
  remaining_months: number
  percentage: number
}

export interface ProgressUpdateResponse {
  id: string
  user_id: string
  roadmap_id: string
  month_name: string
  completed: boolean
  completed_at: string | null
  created_at: string
  updated_at: string
}

export const progressApi = {
  getByRoadmap: (roadmapId: string) =>
    api.get<ProgressSummary>('/progress', { params: { roadmap_id: roadmapId } }),

  update: (data: { roadmap_id: string; month_name: string; completed: boolean }) =>
    api.post<ProgressUpdateResponse>('/progress/update', data),
}
